### PR TITLE
Rest API query and form parameters

### PR DESCRIPTION
**Pull request by mkrebs81**
Support for HTTP POST requests with form data (content type = application/x-www-form-urlencoded).

Separates request query and form parameters from request path or body (with EID specification changes).

This should also solve potential URL encoding problems.

### CODE REVIEW

File: InterfaceFactory/CommHandler4Modbus/src/test/java/communicator/rest/http/client/ApacheRestServiceClientTest.java [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+import static org.junit.jupiter.api.Assertions.assertNotNull;
+import static org.junit.jupiter.api.Assertions.assertNull;
+
+class ApacheRestServiceClientTest {
+
+    @Test
+	void testFormDataFromBodySuccessNoParams() {
+
+        final String body = "";
+        final Iterable<NameValuePair> params = ApacheRestServiceClient.getFormParameters(body);
```
Comment (ergo-furrer) [inline InterfaceFactory/CommHandler4Modbus/src/test/java/communicator/rest/http/client/ApacheRestServiceClientTest.java]: This does not compile (anymore). There is no static method `ApacheRestServiceClient.getFormParameters(body)`

Comment (mkrebs81) [reply InterfaceFactory/CommHandler4Modbus/src/test/java/communicator/rest/http/client/ApacheRestServiceClientTest.java]: I fixed it in my last commit.

The only thing left would be to adapt the _SGrRestAPIDeviceTest_ (I was not sure how to capture the body string)

**Comment by ergo-furrer** [comment]
@mkrebs81:  I will have a look at the`RestApiDeviceTest` and try to fix it.

**Comment by ergo-furrer** [comment]
@mkrebs81 Ifixed the unit-test. Shall I push here? -> Arnd can approve...

**Comment by mkrebs81** [comment]
@ergo-furrer thank you;
yes please push into this branch so Arnd can approve.

**Comment by ergo-furrer** [comment]
@mkrebs81: My test with the real Clemap device does not work anymore. I will have a look at it.

**Comment by mkrebs81** [comment]
@ergo-furrer hmm, maybe I unintentionally changed the path from `/digitaltwins/` to `/digitaltwins`.
Can you check if the missing ending slash in the path is the problem?

**Comment by ergo-furrer** [comment]
> @ergo-furrer hmm, maybe I unintentionally changed the path from `/digitaltwins/` to `/digitaltwins`. Can you check if the missing ending slash in the path is the problem?

Thanks for the suggestion. The authentication fails.  It seems to be an issue with the content-type. I will come back if I got it solved...

**Comment by ergo-furrer** [comment]
> > @ergo-furrer hmm, maybe I unintentionally changed the path from `/digitaltwins/` to `/digitaltwins`. Can you check if the missing ending slash in the path is the problem?
> 
> Thanks for the suggestion. The authentication fails. It seems to be an issue with the content-type. I will come back if I got it solved...

I fixed the `SGr_02_0018_CLEMAP_EIcloudEnergyMonitor_V1.0.0.xml` EI-XML within SGrSpecificaitons.  
- Added `Content-Type: application/xml` header to the 
- Replaced `inline` URL parameter `sensor_id` into a `<requestQuery>` element.

Will do a PR on `SGrSpecifications`

**Comment by ergo-furrer** [comment]
> @ergo-furrer you can commit to my PR which I created today

Arnd asked me to check whether there is  a way to stay backwards compatible with EI-XML so that we still can both `/digitaltwins/?sensor_id={{sensor_id}}` and
```xml
  <requestPath>/digitaltwins</requestPath>
  <requestQuery>
                  <parameter>
                    <name>sensor_id</name>
                    <value>{{sensor_id}}</value>
                  </parameter>
  </requestQuery>
```

**Comment by ergo-furrer** [comment]
> ```
> requestQuery
> ```

The only solution I can think of, is to check if there is a `<requestQuery>` there. If there it is not, then we do not apply URL-encoding.

**Comment by mkrebs81** [comment]
@ergo-furrer I committed an update which should allow backwards compatibility.

**Comment by ergo-furrer** [comment]
> @ergo-furrer I committed an update which should allow backwards compatibility.

Thanks, this works fine!

**Comment by ergo-furrer** [comment]
I added a unit-test that tests the REST-API with form-data with my most recent commit.

**Review by ergo-furrer — APPROVED** [review]

**Review by ArndGerns — APPROVED** [review]

**Pull request merged**